contracts/core/proofAPI/CrossL2EventProver.sol

```diff
@@ -0,0 +1,145 @@
+// SPDX-License-Identifier: Apache-2.0
+/*
+ * Copyright 2024, Polymer Labs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+pragma solidity ^0.8.0;
+
+import {RLPReader} from "optimism/libraries/rlp/RLPReader.sol";
+import {MerkleTrie} from "optimism/libraries/trie/MerkleTrie.sol";
+import {Bytes} from "optimism/libraries/Bytes.sol";
+import {PolymerProofs} from "../../libs/PolymerProofs.sol";
+import {AppStateVerifier} from "../../base/AppStateVerifier.sol";
+import {ICrossEventProver} from "../../interfaces/ICrossEventProver.sol";
+import {Ics23Proof} from "../../interfaces/IProofVerifier.sol";
+import {ISignatureVerifier} from "../../interfaces/ISignatureVerifier.sol";
+import {LightClientType} from "../../interfaces/ILightClient.sol";
+
+contract CrossL2EventProver is AppStateVerifier, ICrossEventProver {
+    LightClientType public constant LIGHT_CLIENT_TYPE = LightClientType.SequencerLightClient; // Stored as a constant
+        // for cheap on-chain use
+
+    string public counterpartyClientID;
+    ISignatureVerifier public immutable verifier;
+
+    // Store peptide apphashes for a given height
+    mapping(uint256 => uint256) public peptideAppHashes;
+
+    error CannotUpdateClientWithDifferentAppHash();
+
+    constructor(ISignatureVerifier verifier_, string memory counterpartyClientID_) {
+        verifier = verifier_;
+        counterpartyClientID = counterpartyClientID_;
+    }
+
+    /**
+     * @inheritdoc ICrossEventProver
+     */
+    function validateReceipt(bytes calldata receiptIndex, bytes calldata receiptRLPEncodedBytes, bytes calldata proof)
```

Comment:
Should these be "memory" if often call from another contract? Same would apply to validateEvent below.

Reply:
Side note, having the proof be plain bytes here is nice if we ever used a zkp version of the CrossL2EventProver contract in the future.